Реализовано задание таймер

JuicyFlash opened the pull request.

JuicyFlash committed:
Реализовано задание таймер

BubuntuClu commented on app/controllers/processed_tests_controller.rb:
```diff
@@ -3,7 +3,11 @@
 
-  def show; end
+  def show
+    if @processed_test.time_to_complete.zero?
```
может быть этот код поместить на 21 строку? чтобы полностью отрабатывала логика завершения теста

BubuntuClu commented on app/models/processed_test.rb:
```diff
@@ -41,6 +51,10 @@ def before_validation_set_first_question
     self.current_question = self.test.questions.first if test.present?
   end
 
+  def before_validation_set_time_to_pass
```
зачем это перекладывать из одного места в другое? дублирование информации получается. бери информацию из теста

JuicyFlash committed:
ProcessedTestsController:
Вынес общий код завершения теста в метод. Теперь код завершения теста выполняется и после прохождения теста и после окончания таймера.
В текущей реализации таймера, если он доходит до 0, сразу выполнится перенаправление на завершение теста через метод show. Если перенести проверку таймера из метода show в update перенаправление произойдёт тоько в момент отпрвки ответов (оставил начальный вариант).

ProcessedTest:
 Убрал дублирование данных - время на прохождение теста теперь берётся из tests.
Начальная идея была сократить количество запросов к БД (к таблице tests), продублировав атрибут время на прохождение в модели processed_test

BubuntuClu commented on app/javascript/utilities/timer.js:
```diff
@@ -0,0 +1,28 @@
+    let timer = setInterval(function () {
+        if (time_to_pass <= 0) {
+            clearInterval(timer)
+            location.reload()
```
>В текущей реализации таймера, если он доходит до 0, сразу выполнится перенаправление на завершение теста через метод show. Если перенести проверку таймера из метода show в update перенаправление произойдёт только в момент отправки ответов (оставил начальный вариант).

пропустил это в первую проверку. какой логический смысл обновлять страницу, если время вышло? только потому что логику ты разместил в метод `show`?
если сабмитить форму, то ситуация может быть такая что просто не успел нажать на кнопку, а ответ выбрал. таким образом ответ будет зачтен
делаешь тут не перезагрузку а сабмит формы, и никакой проблемы не происходит с методами `show/update`

BubuntuClu commented on app/models/processed_test.rb:
```diff
@@ -31,6 +31,7 @@ def percent_result
   end
 
   def time_to_complete
+    time_to_pass = self.test.time_to_pass
```
ты забыл удалить колонку из таблицы, лишнюю миграцию и метод колбэка на выставление значения

JuicyFlash committed:
Удалена миграция
Удалена лишняя колонка из processed_tests
Удалён метод из колбэка
